### PR TITLE
feat(PRI-1321): Create a new runner in core-utils

### PR DESCRIPTION
**Pull request by yankovalera**
Changes
---
- We need a new runner that allows us to chain rules that will run **only** when the first rule runIf = true.
- This will help us to chain rules like the fixedRatesWorkplaces [here](https://github.com/ClipboardHealth/open-shifts/blob/main/curated-shifts/src/modules/shift-offers/pricing-engine/rules/fixed-rates-workplaces.ts#L30), instead of manually calling a rule inside another.
```typescript
const fixedRatesWorkplaces = allIfFirst(fixedBasePay, holidayPay, manualCharge)
``` 
And this will run only the fixedBasePay runIf = true (each subsequent rule will verify it's runIf as well).

## Summary by CodeRabbit

- **New Features**
	- Introduced the `allIfFirst` function to execute rules based on the success of the first rule.
	- Added a suite of unit tests for the `allIfFirst` function to ensure expected behavior under various conditions.

- **Bug Fixes**
	- Improved rule execution flow by ensuring subsequent rules are only processed if the first rule's condition is met.

### CODE REVIEW

File: packages/rules-engine/src/lib/runners/allIfFirst.spec.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  run: (context) => appendOutput(context, 4),
+};
+
+describe("allIfFirst", () => {
+  describe("if", () => {
+    it("returns true if the first rule is true", () => {
+      expect(allIfFirst(testRule2, testRule1).runIf(context.input)).toBe(true);
+    });
+
+    it("returns false if the first rule is false", () => {
```
Comment (therockstorm) [inline packages/rules-engine/src/lib/runners/allIfFirst.spec.ts:47]: Should we add a test for if an empty array is passed?

**Review by therockstorm — APPROVED** [review]

File: packages/rules-engine/src/lib/runners/allIfFirst.spec.ts (hunk 1 above)
Comment (yankovalera) [reply packages/rules-engine/src/lib/runners/allIfFirst.spec.ts:47]: Done, good idea.

File: packages/rules-engine/src/lib/runners/allIfFirst.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * Run all rules that return true for `runIf` only if the first rule returns true.
+ *
+ * @param rules The rules to run.
+ */
+export function allIfFirst<TInput, TOutput, TContext extends RuleContext<TInput, TOutput>>(
+  ...rules: Array<Rule<TInput, TOutput, TContext>>
+): Rule<TInput, TOutput, TContext> {
+  return {
+    runIf: (input) => Boolean(rules[0]?.runIf(input)),
```
Comment (PetricaP) [inline packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: nit: I think we shouldn't have a rule, but a simple predicate based on which the rest of the rules will start.

Comment (PetricaP) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: Example:

`allIf(isSchool, schoolRules)`

Comment (yankovalera) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: @PetricaP This is a new runner to support the previous scenario (e.g. [here](https://github.com/ClipboardHealth/open-shifts/blob/main/curated-shifts/src/modules/shift-offers/pricing-engine/rules/fixed-rates-workplaces.ts#L30))

Comment (PetricaP) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: It could have supported that as well with:
`allIf(isFixedWorkplace, all(basePayRule, basePayMarginRule, holidayRule, manualChargeRule))`

Comment (PetricaP) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: This way we wouldn't need a special rule to be first in the chain.

Comment (yankovalera) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: This slightly changes the runner type, which works on a list of rules. We can come back to this later, but in my use cases, we do need an initial rule, and only if this rule passes do we want to run the others.

Comment (yankovalera) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: @therockstorm, what do you think about this suggestion? 

`allIf(isSchool, schoolRules)`

This is something we could definitely need in the future for the school vertical.

Comment (therockstorm) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: Can't we just have one or the other? The fewer the primitive the better, in my opinion. It seems like  "allIf" would handle both scenarios.

Comment (yankovalera) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: @PetricaP I think a downside of allIf() is traceability, for example, with my approach, I'll write something like:

```
export const fixedPayWorkplacePricingRule: PricingRule = {
  runIf(input): boolean {
    ///
  },
  run(context) {
    return appendOutput(context, {
      meta: {
        reasonCode: RepricingCode.FIXED_PAY,
      },
      amount: fixedPay!,// **it can be 0 if we dont need a special amount here**
    });
  },
};
```

This means we get a rule output for FIXED_PAY (even if the output is 0), while in 

`allIf(isFixedWorkplace, all(basePayRule, basePayMarginRule, holidayRule, manualChargeRule))`

We haven't stored the FIXED_PAY rule output.

Comment (PetricaP) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: We could always include a fixed pay dummy rule if we wanted an output in the `all` so I don't think it's necessarily a downside.

allIf(isFixedWorkplace, fixedPayMarkerRule, basePayRule, basePayMarginRule, holidayRule, manualChargeRule)

I have come up with this problem in other cases as well like for RDA where we needed to duplicate the condition in `runIf` in multiple rules because we couldn't make the condition apply to a set of rules.

I think we'll keep have more and more edge cases where different sets of rules apply so we need something like the `allIf` construct.

To me ignoring the conditions in the other rules in this construct and having a convention that the first rule is somehow special seems hacky, but I don't think it's a big enough problem to worry about so it's not a blocker.

Comment (therockstorm) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: With core libraries, people will use the constructs provided, so we want to make sure to provide constructs that are hard to misuse and that don't result in hacky code. I would say that if `allIf` works for more scenarios, we should go with it instead.

Comment (yankovalera) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: @PetricaP, the other rules runIf are not ignored (we still use the run () when running them); it's just that those rules should run only if the **first** one pass.

I'm happy to change this, but I don't like the idea that this runner has a different structure than the others, which only need a list of rules to run.

Comment (therockstorm) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: Wouldn't the types on the function signature be enough to let people know it's different than the other runners? We could also provide example usage and a TypeDoc comment.

This is likely my last day before paternity leave, so I'll let the two of you come to a decision.

Comment (yankovalera) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: We're in the middle of a new experiment that will give us a better understanding of what we need (allIf, or allIfFirst)

Comment (therockstorm) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: @yankovalera Make sure to update the examples and regenerate the example in the README with `npm run embed`, whatever you decide.

Comment (yankovalera) [reply packages/rules-engine/src/lib/runners/allIfFirst.ts:12]: We've added an allIf [here](https://github.com/ClipboardHealth/open-shifts/blob/main/curated-shifts/src/libs/rules-engine/runner/all-if.ts), which will be ported to core-utils in the next few weeks.

It's been helpful to have this in the new rules.
